src/constructors/circulant.c

```diff
@@ -0,0 +1,102 @@
+/*
+   IGraph library.
+   Copyright (C) 2021  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_constructors.h"
+#include "igraph_interface.h"
+
+/**
+ * \function igraph_circulant
+ * \brief Creates a circulant graph.
+ *
+ * A circulant graph <code>G(n, l)</code> consists of \p n vertices \c v_0, ...,
+ * \c v_(n-1) such that for each \c l_i in the list of offsets \p l, \c v_j is
+ * connected to <code> v_((j + l_i) mod n) </code> for all j.
+ *
+ * </para><para>
+ * The function works with both directed and undirected graphs. Multiple edges are
+ * merged, and self loops are ignored. All offsets are taken modulo \p n, so values greater than n as well as negative offsets are permitted.
+ *
+ * \param graph Pointer to an uninitialized graph object, the result will
+ * be stored here.
+ * \param n Integer, \p n is the number of vertices in the circulant graph. It must
+ * be at least 0.
+ * \param l Integer vector, \p l is a list of the offsets within the circulant graph.
+ * \param directed Boolean, \p directed determines whether the graph should be directed.
+ * \return Error code.
+ *
+ * \sa \ref igraph_ring(), \ref igraph_generalized_petersen().
+ *
+ * Time complexity: O(|V||L|), the number of vertices in the graph times the number
+ * of offsets.
+ */
+
+igraph_error_t igraph_circulant(igraph_t *graph, igraph_integer_t n, const igraph_vector_int_t *l, igraph_bool_t directed) {
+
+    igraph_vector_int_t edges;
+    igraph_vector_bool_t offset_seen;
+    igraph_integer_t i, j;
+    igraph_integer_t limit;
+
+    if (n < 0) {
+        IGRAPH_ERRORF("number of nodes = %" IGRAPH_PRId " must be non-negative.", IGRAPH_EINVAL, n);
+    }
+    if (n == 0) {
+        return igraph_empty(graph, 0, directed);
+    }
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 0);
```

Comment:
Usually, there will be `size(l) * n` edges. Sometimes, there will be fewer because of graph simplification. Here, we _could_ reserve `2 * size(l) * n` amount of space for the edges vector using `IGRAPH_CHECK(igraph_vector_int_reserve(...))`, so that all the `igraph_vector_int_push_back()` calls later will not keep re-allocating.

Ultimately, it's not a big deal here because this is a not a performance-critical function. I still wanted to point this out because in _some_ cases such an optimization can make a noticeable different, and I expect that you will continue contributing @Gomango999, so it's good to be aware of it.

Reply:
Done. I expect to contributing again, so definitely another thing for me to make a mental note of next time.